File: 49. Group Anagrams/49.java
// 49. Group Anagrams solution in Java
// https://leetcode.com/problems/group-anagrams/description/

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List> result_map = new HashMap<String, List>();

        for (String str : strs) {
            int[] key_count = new int[26];

            Arrays.fill(key_count, 0);

            for (char c : str.toCharArray()) {
                key_count[c - 'a']++;
            }

            StringBuilder key_sb = new StringBuilder("");

            for (int key : key_count) {
                key_sb.append(key);
                key_sb.append("#");
            }

            String key = key_sb.toString();

            if (!result_map.containsKey(key)) {
                result_map.put(key, new ArrayList());
            }

            result_map.get(key).add(str);
        }

        return new ArrayList(result_map.values());
    }
}

File: 49. Group Anagrams/49.go
// 49. Group Anagrams solution in GoLang
// https://leetcode.com/problems/group-anagrams/description/

func groupAnagrams(strs []string) [][]string {
    result_map := map[[26]int][]string{}

    for _, str := range strs {
        var key [26]int

        for _, char := range str {
            key[char - 'a']++
        }

        result_map[key] = append(result_map[key], str)
    }

    var result [][]string

    for _, result_map_row := range result_map {
        result = append(result, result_map_row)
    }

    return result
}